src/detect-dns-name.c

```diff
@@ -0,0 +1,188 @@
+/* Copyright (C) 2025 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \file
+ *
+ * Detect keyword for DNS rrnames:
+ * - dns.queries.rrname
+ * - dns.answers.rrname
+ * - dns.authorities.name
+ * - dns.additionals.name
+ */
+
+#include "detect.h"
+#include "detect-parse.h"
+#include "detect-engine.h"
+#include "detect-engine-content-inspection.h"
+#include "detect-engine-helper.h"
+#include "detect-dns-name.h"
+#include "rust.h"
+
+enum {
+    DNS_QUERY = 0,
+    DNS_ANSWER,
+    DNS_AUTHORITY,
+    DNS_ADDITIONAL,
+};
+
+static int query_buffer_id = 0;
+static int answer_buffer_id = 0;
+static int authority_buffer_id = 0;
+static int additional_buffer_id = 0;
+
+static int DetectSetup(DetectEngineCtx *de_ctx, Signature *s, const char *str, int id)
+{
+    if (DetectBufferSetActiveList(de_ctx, s, id) < 0) {
+        return -1;
+    }
+    if (DetectSignatureSetAppProto(s, ALPROTO_DNS) < 0) {
+        return -1;
+    }
+
+    return 0;
+}
+
+static int SetupQueryBuffer(DetectEngineCtx *de_ctx, Signature *s, const char *str)
+{
+    return DetectSetup(de_ctx, s, str, query_buffer_id);
+}
+
+static int SetupAnswerBuffer(DetectEngineCtx *de_ctx, Signature *s, const char *str)
+{
+    return DetectSetup(de_ctx, s, str, answer_buffer_id);
+}
+
+static int SetupAdditionalsBuffer(DetectEngineCtx *de_ctx, Signature *s, const char *str)
+{
+    return DetectSetup(de_ctx, s, str, additional_buffer_id);
+}
+
+static int SetupAuthoritiesBuffer(DetectEngineCtx *de_ctx, Signature *s, const char *str)
+{
+    return DetectSetup(de_ctx, s, str, authority_buffer_id);
+}
+
+static InspectionBuffer *GetBuffer(DetectEngineThreadCtx *det_ctx,
+        const DetectEngineTransforms *transforms, Flow *f, uint8_t flags, void *txv, int list_id,
+        uint32_t index, int what)
```

Comment:
I think it would be nice to make `what` here use the enum type above, helps devs and the compiler

Reply:
Done.